keras_nlp/metrics/bleu.py

```diff
@@ -0,0 +1,385 @@
+
+        if not tf.as_dtype(self.dtype).is_floating:
+            raise ValueError(
+                "`dtype` must be a floating point type. "
+                f"Received: dtype={dtype}"
+            )
+
+        def default_tokenizer(inputs):
+            """
+            Default tokenizer. Replicates the behaviour of SacreBLEU's
+            default tokenizer, namely, `tokenizer_13a`.
+            """
+            for pattern, replacement in REPLACE_SUBSTRINGS + REGEX_PATTERNS:
+                inputs = tf.strings.regex_replace(
+                    input=inputs,
+                    pattern=pattern,
+                    rewrite=replacement,
+                    replace_global=True,
+                    name=None,
+                )
+            inputs = tf.strings.split(inputs)
+            return inputs
+
+        if tokenizer is None:
+            self.tokenizer = default_tokenizer
+        else:
+            self.tokenizer = tokenizer
+        self.max_order = max_order
+        self.smooth = smooth
+
+        self._matches = self.add_weight(
+            shape=(self.max_order,),
+            name="bleu_matches",
+            initializer="zeros",
+            dtype=self.dtype,
+        )
+        self._possible_matches = self.add_weight(
+            shape=(self.max_order,),
+            name="bleu_possible_matches",
+            initializer="zeros",
+            dtype=self.dtype,
+        )
+        self._translation_length = self.add_weight(
+            name="bleu_translation_length",
+            initializer="zeros",
+            dtype=self.dtype,
+        )
+        self._reference_length = self.add_weight(
+            name="bleu_reference_length",
+            initializer="zeros",
+            dtype=self.dtype,
+        )
+        self._bleu = self.add_weight(
+            name="bleu",
+            initializer="zeros",
+            dtype=self.dtype,
+        )
+
+    def _get_ngrams(self, segment, max_order):
+        """Extracts all n-grams upto a given maximum order from an input segment.
+
+        Uses Python ops. Inspired from
+        https://github.com/tensorflow/nmt/blob/master/nmt/scripts/bleu.py.
+
+        Args:
+            segment: string. Text segment from which n-grams will be
+                extracted.
+            max_order: int. Maximum length in tokens of the n-grams returned
+                by this methods.
+        """
+        ngram_counts = collections.Counter()
+        for order in range(1, max_order + 1):
+            for i in range(0, len(segment) - order + 1):
+                ngram = tuple(segment[i : i + order])
+                ngram_counts[ngram] += 1
+        return ngram_counts
+
+    def _corpus_bleu(
+        self,
+        reference_corpus,
+        translation_corpus,
+        matches_by_order,
+        possible_matches_by_order,
+        translation_length,
+        reference_length,
+        max_order=4,
+        smooth=False,
+    ):
+        """Corpus BLEU implementation using Python ops.
```

Comment:
@mattdangerw Matt do we want formal docstring in private method? This method is complicated, so having a detailed explanation looks okay to me. But our general guideline is not to have it?

Reply:
I think it's not necessary, but fine to have if we think it adds something.  In this case it does.